// cloudinary-react/lib/Util/Util.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isElement = exports.cancelAnimationFrame = exports.requestAnimationFrame = exports.closestAbove = exports.firstDefined = exports.debounce = undefined;

var _requestAnimationFrame = require('./requestAnimationFrame');

Object.defineProperty(exports, 'requestAnimationFrame', {
  enumerable: true,
  get: function get() {
    return _requestAnimationFrame.requestAnimationFrame;
  }
});
Object.defineProperty(exports, 'cancelAnimationFrame', {
  enumerable: true,
  get: function get() {
    return _requestAnimationFrame.cancelAnimationFrame;
  }
});

var _debounce2 = require('./debounce');

var _debounce3 = _interopRequireDefault(_debounce2);

var _firstDefined2 = require('./firstDefined');

var _firstDefined3 = _interopRequireDefault(_firstDefined2);

var _closestAbove2 = require('./closestAbove');

var _closestAbove3 = _interopRequireDefault(_closestAbove2);

var _isElement2 = require('./isElement');

var _isElement3 = _interopRequireDefault(_isElement2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.debounce = _debounce3.default;
exports.firstDefined = _firstDefined3.default;
exports.closestAbove = _closestAbove3.default;
exports.isElement = _isElement3.default;
